impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java

```diff
@@ -49,138 +63,134 @@ public class KafkaClusterTest {
     private static final Duration CLUSTER_FORMATION_TIMEOUT = Duration.ofSeconds(10);
     private TestInfo testInfo;
     private KeytoolCertificateGenerator keytoolCertificateGenerator;
+    private static int clusterSizesIndex = 0;
 
-    @Test
-    public void kafkaClusterKraftMode() throws Exception {
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .kraftMode(true)
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(1, cluster);
-        }
+    private static Stream<Version> versions() {
+        return Stream.of(
+                version("latest")
+        // TODO: waiting for new versions support in ozangunalp repo https://github.com/ozangunalp/kafka-native/issues/21
+        // version("3.3.1"),
+        // version("3.2.1")
+        );
     }
 
-    @Test
-    public void kafkaClusterZookeeperMode() throws Exception {
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .kraftMode(false)
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(1, cluster);
-        }
+    private static Stream<BrokerCluster> clusterSizes() {
+        return Stream.of(
+                brokerCluster(1),
+                brokerCluster(2));
     }
 
-    @Test
-    public void kafkaTwoNodeClusterKraftMode() throws Exception {
-        int brokersNum = 2;
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .brokersNum(brokersNum)
-                .kraftMode(true)
-                .build())) {
-            assumeTrue(cluster instanceof TestcontainersKafkaCluster, "KAFKA-14287: kraft timing out on shutdown in multinode case");
-            cluster.start();
-            verifyRecordRoundTrip(brokersNum, cluster);
-        }
+    private final Object[] brokerClusters = clusterSizes().toArray();
+
+    static Stream<List<Annotation>> tuples() {
+        return Stream.of(
+                List.of(brokerCluster(1), kraftCluster(1)),
+                List.of(brokerCluster(2), kraftCluster(1)),
+                List.of(brokerCluster(1), zooKeeperCluster()),
+                List.of(brokerCluster(2), zooKeeperCluster()));
     }
 
-    @Test
-    public void kafkaTwoNodeClusterZookeeperMode() throws Exception {
-        int brokersNum = 2;
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .brokersNum(brokersNum)
-                .kraftMode(false)
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(brokersNum, cluster);
+    @BeforeEach
+    public void initializeIndexes() {
+        if (clusterSizesIndex == clusterSizes().count()) {
```

Comment:
Why do we have an `if` guard here? Shouldn't it just reset it to zero for each test?

Reply:
No, templates does not need to reset the indexes unless we have covered the total amount of `clusterSizes` dimension. It can be a little bit confusing but it is the only way I found to be able to get the values from `clusterSizes` to be checked inside the test.

Reply:
I'm struggling with the stateful nature that this gives for the tests. 

I'm starting to think the `KafkaCluster` should expose the broker count so that the tests can know how many brokers are involved.

Reply:
AFAIK if you run the test you can see the name of the tests telling the number of brokers used and all the info indicated in the `@DimensionMethodSource`

Reply:
@franvila it's not exactly obvious how to test that a `@TestTemplate` -annotated test is generating the set of tests you expect, but I did something a bit similar in `io.kroxylicious.testing.kafka.junit5ext.TemplateTest` which I think is a little cleaner:

* When a test is executed it adds to a map
* I do the assertion in an `@AfterAll`

It's not ideal, but I couldn't think of a nicer way without executing a test runner with an `@Test` which I felt was more mind bending.

Btw I still think that the `@TestTemplate` should be tested in `io.kroxylicious.testing.kafka.junit5ext.TemplateTest` rather than `KafkaClusterTest`. i.e. I still don't really understand why you've moved all these classes to `common`. In my mind the layering is:

```
common
  ^
  |
impl
  ^
  |
junit5
```

And that's because common and impl provide a useful API that doesn't depend on junit in any way. E.g. _in principal_ we could use the same classes to provide some similar functionality for an extension to some non-JUnit testing framework.

Reply:
@tombentley I saw your tests in `io.kroxylicious.testing.kafka.junit5ext.TemplateTest` but the intention here is not testing the template itself, it is to make use of the template as a tool to perform an integration test and then in each test check if the kafka with the corresponding configuration is working fine or not. That's why I keep it in `KafkaClusterTest` class.

I have moved all classes to common because if I hadn't do that, we would have a circle dependency between `junit5` and `common`, that is I cannot see what is defined in `junit5` from `common` unless I add the dependency in the pom file what makes the circle dependency.